### PR TITLE
Panic mode

### PR DESCRIPTION
**Pull request by marcSoda**
Add basic panic mode recovery to parser.

### CODE REVIEW

File: crates/lightc/src/main.rs [hunk 1]
```diff
@@ -103,7 +106,7 @@ fn main() {
         .arg("-o")
         .arg(outfile)
         .arg(module_file)
-        .arg("stdlib/stdlib.o")
+        // .arg("stdlib/stdlib.o")
```
Comment (marcantonio) [inline crates/lightc/src/main.rs:109]: Remove this please.

File: crates/parse/src/lib.rs [hunk 2]
```diff
@@ -32,13 +32,20 @@ impl<'a> Parse<'a> {
...
+            match self.parse_stmt() {
+                Ok(n) => if errs.is_empty() { ast.add(n) },
+                Err(e) => {
+                    errs.push(e);
+                    self.recover();
+                },
+            };
         }
-        Ok(ast)
+        if !errs.is_empty() { Err(errs) }
```
Comment (marcantonio) [inline crates/parse/src/lib.rs:47]: rustfmt

File: crates/parse/src/lib.rs [hunk 3]
```diff
@@ -645,4 +652,25 @@ impl<'a> Parse<'a> {
         }
         Ok(args)
     }
+
+    //Move token iter past next panic_stop_token
+    fn recover(&mut self) {
+        while self.tokens.peek().is_some() {
```
Comment (marcantonio) [inline crates/parse/src/lib.rs:658]: Use `while let`.

**Review by marcantonio — CHANGES_REQUESTED** [review]
Add some test cases.

File: crates/parse/src/lib.rs [hunk 4]
```diff
@@ -22,23 +22,31 @@ type ParseResult = Result<ast::Node, ParseError>;
 pub struct Parse<'a> {
     tokens: Peekable<Iter<'a, Token>>,
     symbol_table: &'a mut SymbolTable<Symbol>,
+    errs: Vec<ParseError>,
```
Comment (marcantonio) [inline crates/parse/src/lib.rs:25]: nit: Can you not abbreviate this?

File: crates/parse/src/lib.rs [hunk 5]
```diff
@@ -22,23 +22,31 @@ type ParseResult = Result<ast::Node, ParseError>;
...
-    pub fn parse(mut self) -> Result<Ast<ast::Node>, ParseError> {
+    pub fn parse(mut self) -> Result<Ast<ast::Node>, Vec<ParseError>> {
         let mut ast = Ast::new();
         while self.tokens.peek().is_some() {
-            let node = self.parse_stmt()?;
-            ast.add(node);
+            match self.parse_stmt() {
+                Ok(n) if self.errs.is_empty() => ast.add(n),
+                Err(e) => self.push_err(e),
+                _ => continue
```
Comment (marcantonio) [inline crates/parse/src/lib.rs:42]: This is unreachable, right? If so change `continue` to `unreachable!("why this is unreachable...")`

Comment (marcantonio) [reply crates/parse/src/lib.rs:42]: Actually, why not just skip the guard? Will it ever `parse_stmt()` ever return `Ok()` and `errs` be empty?

**Review by marcantonio — CHANGES_REQUESTED** [review]
This is great. A few minor things noted...

File: crates/parse/src/lib.rs (hunk 5 above)
Comment (marcSoda) [reply crates/parse/src/lib.rs:42]: After the parse has successfully recovered, every seccessfully parsed stmt will return Ok(n) with self.errs being non-empty. The last guard is not unreachable and cannot be skipped.

**Review by marcantonio — APPROVED** [review]

**Pull request merged**